README.md
# WCLPerformance

CHANGELOG.md
版本 0.2.0

- 增加报告功能
- 国服部落可加入团队在线更新
- 放弃Skada支持，仅使用Details

Core.lua
local addon_name, Addon = ...

WCLPerf = LibStub("AceAddon-3.0"):NewAddon("WCLPerformance")
Addon.addon = WCLPerf
local Core = WCLPerf
Core.addon_name = "WCLPerformance"
Core.title = GetAddOnMetadata(Core.addon_name, "Title")
Core.version = GetAddOnMetadata(Core.addon_name, "Version")

function Core:OnInitialize()
	local db_defaults = {
		global = {
			update = {}
		},
		profile = {
			config = {
				show_left = false,
				show_right = true,
				color = true,
				bracket = true,
				only_role_spec = true,
			},
			report = {
				disable = false,
				raid_channel = true,
			},
			update = {
				disable = false,
			},
		}
	}
	self.db = LibStub("AceDB-3.0"):New("WCLPerformance_DB", db_defaults,"default")
end


local function _secure_call (...)
	local success, result, result2 = pcall(...)
	if success then return result 
	-- else print(result,result2) 
	end
end


-- function Core:LoadDB()
-- 	if WCLPerf_Database then
-- 		self.db = WCLPerf_Database
-- 	end
-- 	return self.db
-- end

Core.zone_to_addon = {
	[24] = "WCLPerformance_Database_Nyalotha",
}

Core.addon_to_zone = {
	["WCLPerformance_Database_Nyalotha"] = {24},
}


Core.zone_db = {}

Core.encounter_to_zone = {
	[2329] = 24, 
	[2327] = 24, 
	[2334] = 24, 
	[2328] = 24, 
	[2333] = 24, 
	[2335] = 24, 
	[2343] = 24, 
	[2336] = 24, 
	[2331] = 24, 
	[2345] = 24, 
	[2337] = 24, 
	[2344] = 24,
}


function Core:GetPerf_Raw(player_amount, metric, encounter, difficulty, class_id, spec)
	local player_amount = player_amount * 100
	local percentiles = {100,99,95,90,80,75,70,60,50,40,30,25,20,10,0}
	local zoneID = self.encounter_to_zone[encounter]
	local db = self:GetZoneDB(zoneID)
	if not db then return nil end
	local has_spec = self:GetSpecPage(metric, encounter, difficulty, class_id, spec)
	if not has_spec then return nil end
	local up, down, up_amount, down_amount
	for _,percentile in ipairs(percentiles) do
		amount = self:GetPercentileAmount(percentile, metric, encounter, difficulty, class_id, spec)
		if amount then
			if player_amount > amount then
				down = percentile
				down_amount = amount
				break
			else
				up = percentile
				up_amount = amount
			end
		-- else
		-- 	print("no amount", percentile, metric, encounter, difficulty, class_id, spec)
		end
	end

	if not up then 
		return 100
	elseif not down then 
		return up 
	else
		return math.floor(up - (up_amount - player_amount) / (up_amount - down_amount) * (up - down) + 0.5) 
	end
end

function Core:GetZoneDB(zoneID)
	if not zoneID then return nil end
	if self.zone_db[zoneID] == false then 
		return nil
	elseif self.zone_db[zoneID] then
		return self.zone_db[zoneID] 
	end
	-- if nil
	return self:ZoneDBRefresh(zoneID)
end


function Core:ZoneDBRefresh(zoneID)
	local db_version = self.db.global.update.date

	local addon_version
	local addon_name = Core.zone_to_addon[zoneID]
	if addon_name then addon_version = self:GetAddonVersion(addon_name) end

	local db_source
	if db_version and addon_version then
		if db_version >= addon_version then db_source = "db" else db_source = "addon" end
	else 
		if db_version then db_source = "db" end
		if addon_version then db_source = "addon" end
	end

	if db_source == "db" then
		self.zone_db[zoneID] = self.db.global.update.data[zoneID]
	elseif db_source == "addon" then
		if not IsAddOnLoaded(addon_name) then LoadAddOn(addon_name) end
		self.zone_db[zoneID] = WCLPerf_Database and WCLPerf_Database[zoneID]
	else
		self.zone_db[zoneID] = false
		return nil
	end
	return self.zone_db[zoneID]
end


function Core:GetAddonVersion( addon_name )
	local success, version = pcall(function () return date("%y%m%d", tonumber(GetAddOnMetadata(addon_name, "X-Update-Date"))) end)
	if success then return version else print(version) end
end

function Core:GetSpecPage(metric, encounter, difficulty, class_id, spec)
	local zoneID = self.encounter_to_zone[encounter]
	local db = self:GetZoneDB(zoneID)
	if not db then return nil end
	if zoneID == 24 then	--maybe different structure for other raids
		local found, data = pcall(function () return db.data[metric][encounter][difficulty][spec] end)
		if found then return data else return nil end
	end
end


function Core:GetPercentileAmount(percentile, metric, encounter, difficulty, class_id, spec)
	local zoneID = self.encounter_to_zone[encounter]
	local db = self:GetZoneDB(zoneID)
	if not db then return nil end
	if zoneID == 24 then	--maybe different structure for other raids
		local found, amount = pcall(function () return db.data[metric][encounter][difficulty][spec][percentile] end)
		if found then return amount else return nil end
	end
end



function Core:GetPerf(...)
	return _secure_call(self.GetPerf_Raw, self, ...)
end

function Core:GetDatebaseVersion()
	for date,_ in pairs(WCLPerf_Database) do
		return date
	end
end

function Core:PerfColor(perf)
	if perf >= 99 then return ITEM_QUALITY_COLORS[6].hex --Artifact
	elseif perf >= 90 then return ITEM_QUALITY_COLORS[5].hex --Legendary
	elseif perf >= 75 then return ITEM_QUALITY_COLORS[4].hex --Epic
	elseif perf >= 50 then return ITEM_QUALITY_COLORS[3].hex --Rare
	elseif perf >= 25 then return ITEM_QUALITY_COLORS[2].hex --Uncommon
	else return	ITEM_QUALITY_COLORS[0].hex --Poor
	end
end

function Core:GetColoredPerf(...)
	local perf = self:GetPerf(...)
	if perf then return self:PerfColor(perf) .. perf .. '|r' end
end

Core.class_id_list = {
		WARRIOR = 1,
		PALADIN = 2,
		HUNTER = 3, 
		ROGUE = 4, 
		PRIEST = 5, 
		DEATHKNIGHT = 6, 
		SHAMAN = 7, 
		MAGE = 8, 
		WARLOCK = 9, 
		MONK = 10, 
		DRUID = 11, 
		DEMONHUNTER = 12
}

function Core:ClassFileToID(classFile)
	return self.class_id_list[classFile] or 0
end

Core.healer_spec = {[105] = true, [65] = true, [256] = true, [257] = true, [264] = true, [270] = true,}

function Core:IsSpecRole(spec, metric, force)
	local only_role_spec = self.db.profile.config.only_role_spec
	if not only_role_spec and not force then return true end
    if (metric == 'dps' and Core.healer_spec[spec]) or (metric == 'hps' and not Core.healer_spec[spec]) then return false end 
    return true
end

function Core:DetailsText_Raw(...)
	local player, combat, instance = ...; 
    local player_amount, metric, encounter, difficulty, class, spec
    if not player:IsGroupPlayer() then return "" end
    if instance.atributo and instance.atributo == 1 and instance.sub_atributo and (instance.sub_atributo == 1 or instance.sub_atributo == 2) then  --dps
        metric = 'dps'
        player_amount = math.floor(player.last_dps)
    elseif instance.atributo and instance.atributo == 2 and instance.sub_atributo and (instance.sub_atributo == 1 or instance.sub_atributo == 2) then  --hps
        metric = 'hps'
        player_amount = math.floor(player.last_hps)
    else
        return ""
    end
    if combat.is_boss then
        encounter = combat.is_boss.id
        difficulty = combat.is_boss.diff
    else
        return ""
    end
    local guid = player:guid() or player.serial
    class = player:class()
    local class_id = self:ClassFileToID(class)
    spec = player.spec or Details:GetSpec(guid)
    if not (class and spec) then return "" end
    if not self:IsSpecRole(spec, metric) then return "" end
    -- encounter = 2333
    -- difficulty = 16
    -- print("DetailsText_Raw", player_amount, metric, encounter, difficulty, class_id, spec)
    local text 
    local config = self.db.profile.config
    if config.color then
    	text = self:GetColoredPerf(player_amount, metric, encounter, difficulty, class_id, spec)
    else
    	text = self:GetPerf(player_amount, metric, encounter, difficulty, class_id, spec)
    end
    if config.bracket then
    	text = text and '[' .. text .. ']'
    end
    return text or ""
end

function Core:DetailsText(...)
	return _secure_call(self.DetailsText_Raw, self, ...) or ""
end

-- local frame = CreateFrame("Frame")
-- frame:RegisterEvent("PLAYER_ENTERING_WORLD")
-- frame:RegisterEvent("ENCOUNTER_START")
-- frame:SetScript("OnEvent", function(self, event, addon_name)
-- 	self:UnregisterEvent("PLAYER_ENTERING_WORLD")
-- 	if Details then Details:TrackSpecsNow (true) end
-- end)

local frame = CreateFrame("Frame")
frame:RegisterEvent("ADDON_LOADED")
frame:SetScript("OnEvent", function(self, event, addon_name)
	if event == "ADDON_LOADED" then
		local zones = Core.addon_to_zone[addon_name]
		if zones then
			Core:CheckDataBaseUpdate(addon_name)
		end
	end
end)

function Core:CheckDataBaseUpdate(addon_name)
	local server_time = GetServerTime()
	local reminder_time = tonumber(GetAddOnMetadata(addon_name, "X-Update-Date-Reminder"))
	if reminder_time and reminder_time < server_time then
		local update_time = date("%Y-%m-%d %H:%M:%S", tonumber(GetAddOnMetadata(addon_name, "X-Update-Date")))
		print(string.format("%s数据库更新时间：%s, 建议进行更新！", addon_name, update_time))
	end
end


Report.lua
local addon_name, Addon = ...
local Core = Addon.addon
local Report = Core:NewModule("Report")
Core.Report = Report
Report.addon_name = Core.addon_name

-- local AceConfigDialog = LibStub("AceConfigDialog-3.0")

local _,i,j,k,v

function Report:OnInitialize()
	self.db = Core.db.profile.report
end


function Report:GenerateReport(combat, db)
	--check db available
	local difficulty = combat:GetDifficulty()
	--difficulty = 16
	local bossInfo = combat:GetBossInfo()
		--table members: name, zone, mapid, diff, diff_string, id, ej_instance_id, killed, index@
	local encounter = bossInfo.id
	local found, data = pcall(function () return db.data['dps'][encounter][difficulty] or db.data['hps'][encounter][difficulty] end)
	if not (found and data) then return nil end
		
	--generate dps
	local report_dps = {}
	--dps DETAILS_ATTRIBUTE_DAMAGE DETAILS_SUBATTRIBUTE_DPS
	local actorContainer = combat:GetContainer(DETAILS_ATTRIBUTE_DAMAGE)
	actorContainer:SortByKey("total")
	local metric = 'dps'
	for _, player in actorContainer:ListActors() do
	    if (player:IsPlayer()) then
	    	local player_report = self:GenerateReport_Player(player, combat, metric)
	    	if player_report then tinsert(report_dps, player_report) end
	    end
	end
	--generate hps
	local report_hps = {}
	--hps DETAILS_ATTRIBUTE_HEAL DETAILS_SUBATTRIBUTE_HPS
	actorContainer = combat:GetContainer(DETAILS_ATTRIBUTE_HEAL)
	actorContainer:SortByKey("total")
	metric = 'hps'
	for _, player in actorContainer:ListActors() do
	    if (player:IsPlayer()) then
	    	local player_report = self:GenerateReport_Player(player, combat, metric)
	    	if player_report then tinsert(report_hps, player_report) end
	    end
	end

	if #report_dps == 0 and #report_hps == 0 then return nil end

	--generate prefix
	local report = {}
	local combatTime = combat:GetCombatTime()
	local formatedCombatTime = combat:GetFormatedCombatTime()

	report_dps = self:ConcateReport(report_dps)
	report_hps = self:ConcateReport(report_hps)

	local report = {}
	tinsert(report, '-----WCL战斗报告-----')
	tinsert(report, '战斗名称：' .. bossInfo.name .. '[' .. bossInfo.diff_string .. ']')
	tinsert(report, '战斗时长：' .. date("%M:%S", combatTime))
	if db.update then tinsert(report, '评分数据库更新：' .. date("%Y-%m-%d", db.update)) end
	tinsert(report, '-----DPS评分-----')
	for _, line in ipairs(report_dps) do tinsert(report, line) end
	tinsert(report, '-----治疗评分-----')
	for _, line in ipairs(report_hps) do tinsert(report, line) end

	return report

end
--[[
WCL战斗报告:BOSS 
战斗时长：	比95%的团队更快！
数据库更新时间：
DPS评分: [惩戒98]XXXXX， [火焰94]XXXXX， [暗影87]XXXXX，  
HPS评分: [98神圣]XXXXX， [66恢复] XXXXX
]]


function Report:ConcateReport(report)
	local concated_report = {}
	local len = 0
	local temp = {}
	for _, line in ipairs(report) do
		if len + #line > 200 then
			tinsert(concated_report, strjoin("， ", unpack(temp)))
			wipe(temp)
			len = 0
		end
		len = len + #line
		tinsert(temp, line)
	end
	tinsert(concated_report, strjoin("， ", unpack(temp)))
	return concated_report
end


function Report:GenerateReport_Player(player, combat, metric)
	local guid = player:guid() or player.serial
	local spec = player.spec or Details:GetSpec(guid)
	if not spec then return nil end
    if not Core:IsSpecRole(spec, metric, true) then return nil end
    local player_amount
    if metric == 'dps' then
    	player_amount = math.floor(player.last_dps)
    elseif metric == 'hps' then
    	player_amount = math.floor(player.last_hps)
    end
	local difficulty = combat:GetDifficulty()
	-- difficulty = 16
	local bossInfo = combat:GetBossInfo()
		--table members: name, zone, mapid, diff, diff_string, id, ej_instance_id, killed, index@
	local encounter = bossInfo.id
	
	local performance = Core:GetPerf(player_amount, metric, encounter, difficulty, class_id, spec)
	if not performance then return nil end

	--spec_name = select(2, GetSpecializationInfoByID(spec))
	player_name = player:GetOnlyName()

	if player_name and performance then
		return '['..performance..'分]'..player_name
	end
	return nil
end


function Report:UseRaidChannel(combat, db)
	if not self.db.raid_channel then return false end
	local difficulty = combat:GetDifficulty()
	if not difficulty or difficulty ~= 16 then return false end
	local update = db.update_time  --timestamp
	if not update then return false end
	update = tonumber(update)
	local server_time = GetServerTime()
	if server_time > update + 7 * 24 * 3600 then return false end
	return true
end

function Report:SendReport(report, use_raid_channel)
	for _, line in ipairs(report) do
		if use_raid_channel then 
			SendChatMessage(line, 'RAID') 
		else 
			--SendChatMessage(line, 'SAY') 
			DEFAULT_CHAT_FRAME:AddMessage(line) 
		end
	end
end

function Report:ON_BOSS_KILL(event, encounterID)
	-- print(event, encounterID)
	--check disable
	local self = Report
	if self.db.disable then return nil end
	--check encounterID
	local zoneID = Core.encounter_to_zone[encounterID]
	if not zoneID then return nil end

	--check database
	local db = Core:GetZoneDB(zoneID)
	if not db then return nil end
	--get combat info
	local combat = Details:GetCurrentCombat()

	--GenerateReport
	local report = self:GenerateReport(combat, db)
	local use_raid_channel = self:UseRaidChannel(combat, db)
	--SendReport
	if report then self:SendReport(report, use_raid_channel) end
end


local tracker = CreateFrame("Frame")
tracker:RegisterEvent("BOSS_KILL")
tracker:SetScript("OnEvent", Report.ON_BOSS_KILL)
Report.tracker = trakcer

Update.lua
local addon_name, Addon = ...
local Core = Addon.addon
local Update = Core:NewModule("Update")
Core.Update = Update
Update.addon_name = Core.addon_name

local AceConfigDialog = LibStub("AceConfigDialog-3.0")
local AceSerializer = LibStub:GetLibrary("AceSerializer-3.0")
local LibDeflate = LibStub:GetLibrary("LibDeflate")

--Embed AceComm
LibStub:GetLibrary("AceComm-3.0"):Embed(Update)



local _,i,j,k,v

function Update:OnInitialize()
	self.db = Core.db.global.update
	if GetCurrentRegion() ~= 5 then
		self.disable = true
	elseif UnitFactionGroup("player") == "Horde" then --部落
		self.author_sender = "数据更新员-亚雷戈斯"
		self.invite_message = "组我更新WCL数据库"
	else
		self.disable = true
	end
	self:RegisterUpdateChannels()
end

function Update:GetCurrentDate()
	return date("%y%m%d", GetServerTime() - 4 * 3600)	--update at 4:00 am
end

-- function Update:GetCurrentVersion(database)
-- 	local addon_name = Core.zone_to_addon[zoneID]
-- 	if addon_name then addon_version = string.match(GetAddOnMetadata(addon_name, "Version"),"%d+") end
-- end

function Update:IsUpdateNeed()
	if self.disable or Core.db.profile.update.disable then return false end

	local db_version = self.db.date
	if db_version and db_version >= self:GetCurrentDate() then return false end
	local addon_min_version = self:GetAddonMinVersion()
	if addon_min_version and addon_min_version >= self:GetCurrentDate() then return false end

	return true
end


function Update:GetAddonMinVersion()
	local addon_min_version
    for zoneID, _ in pairs(Core.zone_to_addon) do
    	local addon_version
		local addon_name = Core.zone_to_addon[zoneID]
		if addon_name then addon_version = Core:GetAddonVersion(addon_name) end
		if addon_min_version and addon_version then 
			addon_min_version = (addon_min_version > addon_version) and addon_version
		elseif not addon_min_version then 
			addon_min_version = addon_version
		end
    end
    return addon_min_version
end


function Update:RegisterUpdateChannels()
	if self.disable or Core.db.profile.update.disable then
		self:UnregisterAllComm()
		return nil 
	end
	if not self:IsUpdateNeed() then return nil end
	local prefix = self:GetUpdataPrefix()
	self:RegisterComm(prefix, self.UpdateMessageReceived)
end

function Update:GetUpdataPrefix()
	local prefix = "WCLPUR" --WCLP, Update, Raid
	prefix = prefix .. self:GetCurrentDate()
	return prefix
end

function Update.UpdateMessageReceived(prefix, message, distribution, sender)
	local self = Update
	if self:SenderFullname(sender) ~= self.author_sender then return nil end
	if message == "START_UPDATE" then
		print("开始更新WCL数据库，数据传输中，预计2-3分钟完成")
		return nil
	end
    local decoded_string = LibDeflate:DecodeForWoWAddonChannel(message)
    local decompressed_string = LibDeflate:DecompressDeflate(decoded_string)
    local success,message_data =  AceSerializer:Deserialize(decompressed_string)
    if not success or type(message_data) ~= "table" then 
        return nil 
    end
    wipe(self.db)
    self.db.data = message_data.data
    self.db.date = message_data.date
    for zoneID, _ in pairs(Core.zone_to_addon) do
        Core:ZoneDBRefresh(zoneID)
    end
    print("WCL数据库更新完成，版本："..message_data.date)
    -- self:SendCommMessage("WCLPUR", "COMPLETE", distribution, sender)
    C_PartyInfo.LeaveParty()
    self:UnregisterAllComm()
end

function Update:SenderFullname(sender)
	if select(2,strsplit('-', sender)) then 
		return sender
	else
		return sender .. '-' .. GetNormalizedRealmName()
	end
end

-- local tracker = CreateFrame("Frame")
-- tracker:RegisterEvent("PLAYER_ENTERING_WORLD")
-- tracker:SetScript("OnEvent", function(self, event, addon_name)
-- 	Update:RegisterUpdateChannels()
-- end)
-- Update.tracker = trakcer

Options.lua
local addon_name, Addon = ...
local Core = Addon.addon
local Options = Core:NewModule("Options")
Core.Options = Options
Options.addon_name = Core.addon_name

local AceConfigDialog = LibStub("AceConfigDialog-3.0")

local _,i,j,k,v

local community_hidden = GetCurrentRegion() ~= 5

function Options:OnInitialize()
	self.db = Core.db.profile
	self:Options_Create()
	self:Options_Register()
end

function Options:Options_Create()
	self.options = {
	    name = self.addon_name,
	    handler = self,
	    type = 'group',
	    childGroups = "tab",
	    args = {
	    	info = {
	    		type = 'group',
				name = "",
				inline = true,
				order = 1,
				args = {
					title = {
						type = "description",
						name = function () return "|cffffd100" .. select(2,GetAddOnInfo(self.addon_name)) .. "|r" end,
						fontSize = "large",
						order = 1,
						width = 2,
					},
					join = {
						type = "execute",
						hidden = function () return community_hidden end,
						name = "加入阳光插件社区",
						desc = "加入社区，交流使用问题",
						func = "JoinCommunity",
						order = 2,
					},
					author = {
						type = "description",
						name = "|cffffd100Author: |r" .. GetAddOnMetadata(self.addon_name, "Author"),
						order = 4,
						width = 2,
					},
					version = {
						type = "description",
						name = "|cffffd100Version: |r" .. GetAddOnMetadata(self.addon_name, "Version") .."\n",
						order = 5,
					},
				},
	    	},
	    },
	}
	self.options.args.config = self:Options_Create_Config()
	-- self.options.args.skada = self:Options_Create_Skada()
	self.options.args.database = self:Options_Create_Database()
end


--[[
配置 Details
显示位置：统计行左侧 统计行右侧
导入自定义文字	还原
配置显示天赋

显示排名颜色
排名加[]包围

]]
function Options:Options_Create_Config()
	local config = self.db.config
	local report = self.db.report
	local options = {
	    name = "配置选项",
	    handler = self,
	    type = 'group',
	    -- inline = true,
	    disabled = function () if not Details then return true end end,
	    order = 11,
	    args = {
	    	show_setting = {
	    		type = "group",
	    		name = "Details必要设置",
	    		inline = true,
	    		order = 1,
	    		args = {
	    			-- header = {
	    			-- 	type = "header",
	    			-- 	name = "必要设置",
	    			-- 	order = 0,
	    			-- },
	    			show = {
	    				type = "description",
	    				name = "|cffffd1001. 选择显示位置：|r可以左右都显示，左侧刷新略慢一怕",
			    		order = 11,
	    			},
			    	show_left = {
			    		type = "toggle",
			    		name = "左侧显示",
			    		desc = "WCL分数显示在名字前面",
			    		get = function () return config.show_left end,
			    		set = function (info, val) config.show_left = val end,
			    		order = 12,
			    	},
			    	show_right = {
			    		type = "toggle",
			    		name = "右侧显示",
			    		desc = "WCL分数显示在DPS/HPS值后面",
			    		get = function () return config.show_right end,
			    		set = function (info, val) config.show_right = val end,
			    		order = 13,
			    	},
	    			custom_text_info = {
	    				type = "description",
	    				name = "\n|cffffd1002. 设置自定义文字：|r通过Details自定义文字显示WCL分数，更改上方选项之后，需要重新按一次设置按钮",
			    		order = 21,
	    			},
			    	custom_text = {
			    		type = "execute",
			    		name = "设置自定义文字",
			    		desc = "在自定义文字中显示WCL分数",
			    		func = "Details_SetCustomText",
			    		order = 22,
			    	},
			    	custom_text_reset = {
			    		type = "execute",
			    		name = "还原",
			    		desc = "不再显示WCL分数",
			    		func = "Details_SetCustomText_Reset",
			    		width = 0.5,
			    		order = 23,
			    	},
	    			custom_text_info2 = {
	    				type = "description",
	    				name = "也可以将下方文本框内的内容复制粘贴到Details自定义文字中，自行设置显示方式",
			    		order = 24,
	    			},
	    			custom_text_input = {
	    				type = "input",
	    				name = "",
	    				get = function () return [[{func return WCLPerf and WCLPerf:DetailsText(...) or ""}]] end,
	    				width = "full",
			    		order = 25,
	    			},
	    			show_spec_info = {
	    				type = "description",
	    				name = "\n|cffffd1003. 追踪天赋：|rWCL分数是分天赋计算的，必须追踪天赋，Details才会查询和记录天赋信息，才能显示分数。之后请谨慎使用Details自身的图标修改功能，可能会使追踪失效。",
			    		order = 31,
	    			},
			    	show_spec = {
			    		type = "execute",
			    		name = "追踪天赋",
			    		desc = "在统计条前面显示天赋图标",
			    		func = "Details_SetSpecShow",
			    		order = 32,
			    	},
			    	show_spec_hide_icon = {
			    		type = "execute",
			    		name = "隐藏图标",
			    		desc = "追踪仍然生效，仅隐藏图标",
			    		func = "Details_SetSpecShow_HideIcon",
			    		order = 33,
			    	},
			    	show_spec_reset = {
			    		type = "execute",
			    		name = "还原",
			    		desc = "不再显示天赋图标",
			    		func = "Details_SetSpecShow_Reset",
			    		width = 0.5,
			    		order = 34,
			    	},
	    		},
	    	},
	    	custom_setting = {
	    		type = "group",
	    		name = "显示设置",
	    		inline = true,
	    		order = 2,
	    		args = {
	    			-- header = {
	    			-- 	type = "header",
	    			-- 	name = "自定义设置",
	    			-- 	order = 0,
	    			-- },
					color = {
						type = "toggle",
						name = "显示WCL分数颜色",
						desc = "例： |cffe6cc80100|r,|cffff800090|r,|cffa335ee75|r,|cff0070dd50|r,|cff1eff0025|r",
						get = function () return config.color end,
						set = function (info, val) config.color = val end,
						order = 1,
					},
					bracket = {
						type = "toggle",
						name = "数字加[]",
						desc = "例： [|cffff800090|r]",
						get = function () return config.bracket end,
						set = function (info, val) config.bracket = val end,
						order = 2,
					},
					only_role_spec = {
						type = "toggle",
						name = "只显示对应职责",
						desc = "DPS职业不显示治疗评分，治疗职业不显示DPS评分",
						get = function () return config.only_role_spec end,
						set = function (info, val) config.only_role_spec = val end,
						order = 3,
					},
	    		},
	    	},
			report = {
	    		type = "group",
	    		name = "报告设置",
	    		inline = true,
	    		order = 3,
	    		args = {
					disable = {
						type = "toggle",
						name = "战斗结束后自动显示报告",
						desc = "战斗结束后自动显示报告",
						get = function () return not report.disable end,
						set = function (info, val) report.disable = not val end,
						order = 1,
					},
					raid_channel = {
						type = "toggle",
						name = "使用团队频道",
						desc = "战斗报告将发送在团队频道（仅史诗模式，非过期数据）",
						get = function () return report.raid_channel end,
						set = function (info, val) report.raid_channel = val end,
						order = 2,
					},
	    		},
	    	},
	    },
	}
	return options
end

-- function Options:Options_Create_Skada()
-- 	local setting = self.db.skada
-- 	local options = {
-- 	    name = "Skada",
-- 	    handler = self,
-- 	    type = 'group',
-- 	    disabled = function () if not Skada then return true end end,
-- 	    -- inline = true,
-- 	    order = 12,
-- 	    args = {
-- 	    	info = {
-- 				type = "description",
-- 				name = "Skada支持暂未完成，由于Skada不追踪和记录天赋，需要好好研究一下方法",
-- 	    		order = 1,
-- 			},
-- 	    }

-- 	}
-- 	return options
-- end

function Options:Options_Create_Database()
	local update = Core.db.profile.update
	local options = {
	    name = "数据库信息",
	    handler = self,
	    type = 'group',
	    order = 21,
	    args = {
	    	disable_update = {
				type = "toggle",
				name = "禁用在线更新",
				get = function () return update.disable end,
				set = function (info, val) update.disable = val Core.Update:RegisterUpdateChannels() end,
				order = 1,
			},
	    	saved_database = {
				type = "group",
				name = "在线更新数据库（试验中）",
				inline = true,
				disabled = function () return update.disable end,
	    		order = 2,
	    		args = {
	    			version = {
	    				type = "description",
	    				name = function () return self:SavedDatabase_GetVersion() end,
	    				order = 1,
	    			},
	    			update_button = {
	    				type = "execute",
	    				name = "加入团队自动更新",
	    				disabled = function () return not self:SavedDatabase_IsUpdateNeed() end,
	    				func = "SavedDatabase_JoinUpdateGroup",
	    				confirm = function () return self:SavedDatabase_Confirm() end,
	    				order = 2,
	    			},
	    			desc = {
	    				type = "description",
	    				name = "\n|cffffd100在线更新说明：|r\n - 仅支持|cffffd100国服部落|r，更新过程需要2-3分钟；\n - 点击按钮会发消息给我小号，|cffffd100如果在线的话|r，会邀请进入团队；\n - 进入团队后，小号自动发送更新数据，插件会自动接收数据；\n - 接收完成，插件自动退出团队；\n - 在线更新数据由于传输效率较低，仅包括|cffffd100史诗难度、对应职责|r的数据；\n - |cffff0000不要在团队中说话！不要在团队中说话！不要在团队中说话！|r会被加入黑名单！\n",
	    				order = 3,
	    			},
	    		},
			},
	    },
	}
	self:Options_Create_Database_All(options.args)
	return options
end

function Options:Options_Create_Database_All(database_tab)
	for addon_name,_ in pairs(Core.addon_to_zone) do
		database_tab[addon_name] = self:Options_Create_Database_Addon(addon_name)
	end
end

function Options:Options_Create_Database_Addon(addon_name)
	local options = {
		type = "group",
		name = addon_name,
		inline = true,
		args = {
			title = {
				type = "description",
				name = self:GetAddonInstances(addon_name),
	    		order = 1,
			},
			update = {
				type = "description",
				name = "更新时间: " .. date("%Y-%m-%d %H:%M:%S", tonumber(GetAddOnMetadata(addon_name, "X-Update-Date"))),
	    		order = 2,
			},
			status = {
				type = "description",
				name = function () if IsAddOnLoaded(addon_name) then return '状态：|cff00ff00已启用|r' else return '状态：|cffff0000未启用|r' end end,
				width = 1,
	    		order = 3,
			},
			load = {
				type = "execute",
				name = "加载",
				hidden = function () return IsAddOnLoaded(addon_name) end,
				func = function () return LoadAddOn(addon_name) end,
				width = 0.5,
	    		order = 4,
			},
		},
	}
	return options	
end

function Options:Options_Register()
	LibStub("AceConfig-3.0"):RegisterOptionsTable(self.addon_name, self.options)
	AceConfigDialog:AddToBlizOptions(self.addon_name, self.addon_name)
	LibStub("AceConsole-3.0"):RegisterChatCommand("wcl", function() AceConfigDialog:Open(self.addon_name) end)
	AceConfigDialog:SetDefaultSize(self.addon_name, 858, 660)
end

function Options:JoinCommunity()
	local club_link = GetClubTicketLink("eaa30RMfp05", "阳光插件", 0)
	if club_link then
		DEFAULT_CHAT_FRAME:AddMessage("社区链接已生成，请点击 "..club_link)
	end
end

function Options:Details_SetCustomText()
	local config = self.db.config
	local show_left = config.show_left
	local show_right = config.show_right
	local left_text = [[{func return WCLPerf and WCLPerf:DetailsText(...) or ""}{data3}{data2}]]
	local right_text = [[{data1} ({data2}) {func return WCLPerf and WCLPerf:DetailsText(...) or ""}]]

	for _, instance in ipairs (Details.tabela_instancias) do
		--instance:SetBarTextSettings (size, font, fixedcolor, leftcolorbyclass, rightcolorbyclass, leftoutline, rightoutline, customrighttextenabled, customrighttext, percentage_type, showposition, customlefttextenabled, customlefttext, translittest)
		instance:SetBarTextSettings (nil, nil, nil, nil, nil, nil, nil, show_right, right_text, nil, nil, show_left, left_text, nil)
	end
end

function Options:Details_SetCustomText_Reset()
	local show_left = false
	local show_right = false
	local left_text = [[{data1}. {data3}{data2}]]
	local right_text = [[{data1} ({data2}, {data3}%)]]

	for _, instance in ipairs (Details.tabela_instancias) do
		instance:SetBarTextSettings (nil, nil, nil, nil, nil, nil, nil, show_right, right_text, nil, nil, show_left, left_text, nil)
	end
end

function Options:Details_SetSpecShow()
	local iconpath = [[Interface\AddOns\Details\images\spec_icons_normal]]
	for _, instance in ipairs (Details.tabela_instancias) do
		instance:SetBarSettings (nil, nil, nil, nil, nil, nil, nil, nil, iconpath)
		instance:SetBarSpecIconSettings (true, iconpath, true)
	end
	if DetailsOptionsWindow4 then _G.DetailsOptionsWindow4.iconFileEntry:SetText (iconpath) end
	_detalhes:SendOptionsModifiedEvent (DetailsOptionsWindow.instance)
end

function Options:Details_SetSpecShow_HideIcon()
	local iconpath = [[]]
	for _, instance in ipairs (Details.tabela_instancias) do
		instance:SetBarSettings (nil, nil, nil, nil, nil, nil, nil, nil, iconpath)
	end
	if DetailsOptionsWindow4 then _G.DetailsOptionsWindow4.iconFileEntry:SetText (iconpath) end
	_detalhes:SendOptionsModifiedEvent (DetailsOptionsWindow.instance)
end

function Options:Details_SetSpecShow_Reset()
	local iconpath = [[]]
	for _, instance in ipairs (Details.tabela_instancias) do
		instance:SetBarSettings (nil, nil, nil, nil, nil, nil, nil, nil, iconpath)
		instance:SetBarSpecIconSettings (false)
	end
	if DetailsOptionsWindow4 then _G.DetailsOptionsWindow4.iconFileEntry:SetText (iconpath) end
	_detalhes:SendOptionsModifiedEvent (DetailsOptionsWindow.instance)
end

function Options:GetAddonInstances(addon_name)
	local InstanceIDs = {strsplit(",", GetAddOnMetadata(addon_name, "X-InstanceID"))}
	local all = "数据库包含："
	for _, InstanceID in ipairs(InstanceIDs) do
		local link = select(8,EJ_GetInstanceInfo(InstanceID))
		if link then all = all .. link end
	end
	return all
end

function Options:SavedDatabase_GetVersion()
	local db_version = Core.db.global.update.date
	local addon_version = Core.Update:GetAddonMinVersion()
	local today = Core.Update:GetCurrentDate()
	local version_string = ''
	if db_version then
		version_string = version_string .. "在线更新数据库版本"..db_version .. '\n'
	else
		version_string = version_string .. "没有找到在线更新数据库" .. '\n'		
	end
	if addon_version then
		version_string = version_string .. "插件内置数据库版本"..addon_version .. '\n'
	else
		version_string = version_string .. "没有找到插件内置数据库" .. '\n'		
	end
	if Core.Update:IsUpdateNeed() then version_string = version_string .. "|cffff0000需要更新|r" else version_string = version_string .. "|cff00ff00无需更新|r" end
	return version_string
end

function Options:SavedDatabase_IsUpdateNeed()
	return Core.Update:IsUpdateNeed()
end

function Options:SavedDatabase_JoinUpdateGroup()
	local sender = Core.Update.author_sender
	local invite_message = Core.Update.invite_message
	if not (sender and invite_message) then return nil end
	SendChatMessage(invite_message, "WHISPER", nil, sender)
end

function Options:SavedDatabase_Confirm()
	local sender = Core.Update.author_sender
	if not sender then return nil end
	return "即加将入团队，在团队中会自动更新数据库，请注意组队邀请：" .. sender
end
